Add common sensitive names to generated filter parameters

eliotsykes opened the pull request:
### Summary

Common sensitive parameter names to consider adding to generated filter parameters for new apps.

The initial commit in this PR is conservative, only adding `:secret` to the current default `[:password]`.

Further candidates distilled from the `filter_parameters` config from a number of open source Rails projects to consider adding:

```rb
# These will filter exact *and* partial matches,
# e.g. :token would filter: token, api_token, access_token, token_foo
:token,
:_key,
:auth,
:crypt,
:salt,
:certificate,
:otp,
:access,
:private,
:protected,
:birth,
:ssn
```

### Other Information

Similar pull request from 2015 discussing adding to these defaults: https://github.com/rails/rails/pull/19809

The Rails apps searched to find these: https://github.com/eliotsykes/real-world-rails/tree/master/apps

eliotsykes committed:
Add secret to generated filter parameters

GeekOnCoffee commented:
All of those suggestions make sense to me, I'd probably add variants of `birth` as well, but that might be overkill.

eliotsykes commented:
Thanks for the feedback @GeekOnCoffee - `:birth` will cover variations too, filtering covers exact and partial matches here. I've clarified the behaviour in the original summary - thanks!

oniofchaos commented:
Yes, I think this is a great idea, especially for apps that have to deal with compliance.

rafaelfranca commented:
I'm ok with adding most of those things but I'm curious to know why `birth`?

oniofchaos commented:
At least in healthcare, birth is considered PHI.

rafaelfranca commented:
I see. I healthcare applications should add birth in the list. I don't think Rails should have it as default value.

GeekOnCoffee commented:
Birth is actually considered PII, which is potentially a concern in anything financial, not just in healthcare; personally I think the privacy benefits to leaving it out of logs outweigh the benefits to having them in logs.

eliotsykes committed:
Add common sensitive names to generated filter parameters

These added names are distilled from the filter_parameters config of a
number of open source Rails applications.

eliotsykes commented:
Thank you for the feedback. Except for `birth`, all the other sensitive param names have been added in the last commit.

ignatiusreza commented:
I wonder if `ssn` should also be included by default.. it make sense for U.S. citizen, but not for the rest of the world? if we're to include it, should we also consider things like `passport`, `nric`, etc?

eliotsykes commented:
Related, @Sqreen is configured by default to scrub the following. 

@aviat would you happen to know and be able to share how these were arrived at? It'd help inform this change to Rails' default param scrubbing.

> Values that look like they contain credit cards
> - using a basic regular expression: (?:\d[ -]*?){13,16}

> Values of any of the following keys:
> - password
> - secret
> - passwd
> - authorization
> - api_key
> - apikey
> - access_token

Source: https://docs.sqreen.com/guides/how-sqreen-works/#pii-scrubbing

aviat commented:
@eliotsykes we came up with some based on client requests, and we got inspiration from other open source tools. For instance, Sentry is doing something similar: https://github.com/getsentry/raven-ruby/blob/21669e97ee704efe40432ae66a1078cf74030ca6/lib/raven/processor/sanitizedata.rb#L5

eliotsykes commented:
Thanks @aviat 👍  

So the params scrubbed by Sqreen and Sentry that wouldn't be covered by Rails' generated filters (since this PR) are the following:

- `passwd`
- `apikey`
- `social(.*)?sec` (appears equiv to `social(.*)sec`)

If the Rails generated `filter_parameters` is ever revised in the future, these above are candidates to consider covering. The existing filtering patterns could be simplified to cover `passwd` and `apikey`:

- Reduce `:password` to `:pass`
- Reduce `:_key` to `:key`
